content/en/blog/2025/observing-lambdas/index.md

```diff
@@ -0,0 +1,117 @@
+least it should be). You set up a collector which either receives data from your
+application or asks it to provide an up-to-date state of various counters. This
+happens every minute or so, and if it’s a second late or early, no one really
+bats an eye. But what if the application isn’t around for long? What if every
+second waiting for the data to be collected is billed? Then you’re most likely
+thinking of Function-as-a-Service (FaaS) environments, the most well-known being
+AWS Lambda.
+
+In this execution model, functions are called directly, and the environment is
+frozen afterward. You’re only billed for actual execution time and no longer
+need a server to wait for incoming requests. This is also where the term
+serverless comes from. Keeping the function alive until metrics can be collected
+isn’t really an option and even if you were willing to pay for that, different
+invocations will have a completely separate context and not necessarily know
+about all the other executions happening simultaneously. You might now be
+saying: "I'll just push all the data at the end of my execution, no issues
+here!", but that doesn’t solve the issue. You’ll still have to pay for the time
+it takes to send the data and with many invocations, this adds up.
+
+But there is another way! Lambda extension layers allow you to run any process
+alongside your code, sharing the execution runtime and providing additional
+services. With the
+[opentelemetry-lambda](https://github.com/open-telemetry/opentelemetry-lambda/blob/main/collector/README.md)
+extension layer, you get a local endpoint to send data to while it keeps track
+of the Lambda lifecycle and ensures your telemetry gets to the storage layer.
+
+## How does it work?
+
+When your function is called for the first time, the extension layer starts an
+instance of the OpenTelemetry Collector. The Collector build is a stripped down
+version, providing only components necessary in the context of Lambda. It
+registers with the Lambda
+[Extensions API](https://docs.aws.amazon.com/lambda/latest/dg/runtimes-extensions-api.html)
+and
+[Telemetry API](https://docs.aws.amazon.com/lambda/latest/dg/telemetry-api.html).
+By doing this, it receives notifications whenever your function is executed,
+emits a logline, or the execution context is about to be shut down.
+
+### This is where the magic happens
+
+Up until now, this just seems like extra work for nothing. You'll still have to
+wait for the Collector to ship the data, right? This is where the special
+`decouple` processor comes in. It separates the receiving and exporting
+components while interfacing with the Lambda lifecycle. This allows for the
+Lambda to return, even if not all data has been sent. At the next invocation (or
+on shutdown) the Collector continues shipping the data while your function does
+its thing.
+
+{{< figure src="diagram-execution-timing.svg" caption="Diagram showcasing how execution timing differs with and without a Collector">}}
```

Comment:
The legend text at the bottom of the image is difficult to read when viewed in dark mode.  Consider adding a contrasting background box around the legend text.

Reply:
Made some improvements which should improve legibility on both dark and light mode - let me know what you think!

Reply:
much better!